File: 2667.cpp
//
//  2667.cpp
//  
//
//  Created by 우소연 on 2018. 3. 7..
//

#include <stdio.h>
#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;
int a[30][30];
int d[30][30];
int dx[]={0,0,1,-1};
int dy[]={1,-1,0,0};
int n;
int woo[25*25]={0,};
void bfs(int x,int y,int cnt){
    queue<pair<int,int> > q;
    q.push(make_pair(x,y));
    d[x][y]=cnt;
    while(!q.empty()){
        x=q.front().first;
        y=q.front().second;
        q.pop();
        for(int k=0;k<4;k++){
            int nx=x+dx[k];
            int ny=y+dy[k];
            if(nx>=0 && nx<n && ny>=0 && ny<n){
                if(a[nx][ny]==1 && d[nx][ny] == 0){
                    q.push(make_pair(nx,ny));
                    d[nx][ny]=cnt;
                }
            }
        }
    }
}
int main(){
    cin>>n;
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
             scanf("%1d",&a[i][j]);
        }
    }
    int cnt=0;
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            if(a[i][j]==1 && d[i][j]==0){
                bfs(i,j,++cnt);
            }
        }
    }
    cout<<cnt<<'\n';
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            woo[d[i][j]]+=1;
        }
    }
    sort(woo+1, woo+cnt+1);
    for (int i=1; i<=cnt; i++) {
        printf("%d\n",woo[i]);
    }
    return 0;
}

File: 2667.py
from collections import deque
n = int(input())
a = [list(map(int, input())) for _ in range(n)]
c = [[0] * n for _ in range(n)]
q = deque()
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
cnt = []
def bfs(x, y):
    temp = 0
    q.append([x, y])
    c[x][y] = 1
    temp += 1
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < n:
                if c[nx][ny] == 0 and a[nx][ny] == 1:
                    q.append([nx, ny])
                    c[nx][ny] = 1
                    temp += 1
                    # print(temp)
    cnt.append(temp)

for i in range(n):
    for j in range(n):
        if c[i][j] == 0 and a[i][j] == 1:
            bfs(i, j)
cnt.sort()
print(len(cnt))
for k in cnt:
    print(k)

